### Import the file from excel and save to a dataframe

In [107]:
import pandas as pd

df = pd.read_excel (r'C:\Users\Tijesunimi2-PC\Documents\10academy_Telecommunication\Week2_challenge_data_source.xlsx')
df

Bearer Id               Start  Start ms  \
0       13114483460844900352 2019-04-04 12:01:18     770.0   
1       13114483482878900224 2019-04-09 13:04:04     235.0   
2       13114483484080500736 2019-04-09 17:42:11       1.0   
3       13114483485442799616 2019-04-10 00:31:25     486.0   
4       13114483499480700928 2019-04-12 20:10:23     565.0   
...                      ...                 ...       ...   
149996   7277825670196679680 2019-04-29 07:28:42     451.0   
149997   7349883264234609664 2019-04-29 07:28:42     483.0   
149998  13114483573367300096 2019-04-29 07:28:43     283.0   
149999  13114483573367300096 2019-04-29 07:28:43     696.0   
150000                                       NaT       NaN   

                       End  End ms  Dur. (ms)          IMSI  MSISDN/Number  \
0      2019-04-25 14:35:31   662.0  1823652.0  2.082014e+14   3.366496e+10   
1      2019-04-25 08:15:48   606.0  1365104.0  2.082019e+14   3.368185e+10   
2      2019-04-25 11:58:13   652.0  1361762.0  2.082003e+14   3.376063e+10   
3      2019-04-25 07:36:35   171.0  1321509.0  2.082014e+14   3.375034e+10   
4      2019-04-25 10:40:32   954.0  1089009.0  2.082014e+14   3.369980e+10   
...                    ...     ...        ...           ...            ...   
149996 2019-04-30 06:02:33   214.0    81230.0  2.082022e+14   3.365069e+10   
149997 2019-04-30 10:41:33   187.0    97970.0  2.082019e+14   3.366345e+10   
149998 2019-04-30 10:46:12   810.0    98249.0  2.082017e+14   3.362189e+10   
149999 2019-04-30 10:40:34   327.0    97910.0  2.082021e+14   3.361962e+10   
150000                 NaT     NaN        NaN           NaN            NaN   

                IMEI Last Location Name  ...  Youtube DL (Bytes)  \
0       3.552121e+13   9164566995485190  ...        1.585461e+07   
1       3.579401e+13            L77566A  ...        2.024740e+07   
2       3.528151e+13            D42335A  ...        1.972566e+07   
3       3.535661e+13            T21824A  ...        2.138812e+07   
4       3.540701e+13            D88865A  ...        1.525938e+07   
...              ...                ...  ...                 ...   
149996  3.548311e+13            D20434A  ...        1.619167e+07   
149997  3.566051e+13            D10223C  ...        1.387723e+07   
149998  3.572121e+13            T51102A  ...        2.266051e+07   
149999  8.618620e+13            L88342B  ...        8.817106e+06   
150000           NaN                NaN  ...        1.163407e+07   

        Youtube UL (Bytes)  Netflix DL (Bytes)  Netflix UL (Bytes)  \
0             2.501332e+06        8.198936e+06        9.656251e+06   
1             1.911173e+07        1.833841e+07        1.722713e+07   
2             1.469958e+07        1.758779e+07        6.163408e+06   
3             1.514664e+07        1.399465e+07        1.097942e+06   
4             1.896287e+07        1.712458e+07        4.152180e+05   
...                    ...                 ...                 ...   
149996        1.176343e+07        1.788370e+07        1.967816e+07   
149997        8.288284e+06        1.935015e+07        2.129315e+07   
149998        1.855903e+06        9.963942e+06        5.065760e+06   
149999        8.305402e+06        3.322253e+06        1.317259e+07   
150000        1.100941e+07        1.162685e+07        1.100175e+07   

        Gaming DL (Bytes)  Gaming UL (Bytes)  Other DL (Bytes)  \
0            2.780823e+08       1.434415e+07      1.717444e+08   
1            6.087501e+08       1.170709e+06      5.269042e+08   
2            2.295846e+08       3.956300e+05      4.106926e+08   
3            7.995382e+08       1.084972e+07      7.490399e+08   
4            5.277072e+08       3.529801e+06      5.507095e+08   
...                   ...                ...               ...   
149996       5.266097e+08       9.197207e+06      3.264510e+06   
149997       6.268931e+08       4.735033e+06      7.121804e+08   
149998       5.535395e+08       1.339432e+07      1.211009e+08   
149999       3.5253

### Fill all missing values in the dataset with the mean

In [101]:
df.fillna(df.mean())

Bearer Id               Start  Start ms  \
0       13114483460844900352 2019-04-04 12:01:18  770.0000   
1       13114483482878900224 2019-04-09 13:04:04  235.0000   
2       13114483484080500736 2019-04-09 17:42:11    1.0000   
3       13114483485442799616 2019-04-10 00:31:25  486.0000   
4       13114483499480700928 2019-04-12 20:10:23  565.0000   
...                      ...                 ...       ...   
149996   7277825670196679680 2019-04-29 07:28:42  451.0000   
149997   7349883264234609664 2019-04-29 07:28:42  483.0000   
149998  13114483573367300096 2019-04-29 07:28:43  283.0000   
149999  13114483573367300096 2019-04-29 07:28:43  696.0000   
150000                                       NaT  499.1882   

                       End     End ms     Dur. (ms)          IMSI  \
0      2019-04-25 14:35:31  662.00000  1.823652e+06  2.082014e+14   
1      2019-04-25 08:15:48  606.00000  1.365104e+06  2.082019e+14   
2      2019-04-25 11:58:13  652.00000  1.361762e+06  2.082003e+14   
3      2019-04-25 07:36:35  171.00000  1.321509e+06  2.082014e+14   
4      2019-04-25 10:40:32  954.00000  1.089009e+06  2.082014e+14   
...                    ...        ...           ...           ...   
149996 2019-04-30 06:02:33  214.00000  8.123000e+04  2.082022e+14   
149997 2019-04-30 10:41:33  187.00000  9.797000e+04  2.082019e+14   
149998 2019-04-30 10:46:12  810.00000  9.824900e+04  2.082017e+14   
149999 2019-04-30 10:40:34  327.00000  9.791000e+04  2.082021e+14   
150000                 NaT  498.80088  1.046086e+05  2.082016e+14   

        MSISDN/Number          IMEI Last Location Name  ...  \
0        3.366496e+10  3.552121e+13   9164566995485190  ...   
1        3.368185e+10  3.579401e+13            L77566A  ...   
2        3.376063e+10  3.528151e+13            D42335A  ...   
3        3.375034e+10  3.535661e+13            T21824A  ...   
4        3.369980e+10  3.540701e+13            D88865A  ...   
...               ...           ...                ...  ...   
149996   3.365069e+10  3.548311e+13            D20434A  ...   
149997   3.366345e+10  3.566051e+13            D10223C  ...   
149998   3.362189e+10  3.572121e+13            T51102A  ...   
149999   3.361962e+10  8.618620e+13            L88342B  ...   
150000   4.188282e+10  4.847455e+13                NaN  ...   

        Youtube DL (Bytes)  Youtube UL (Bytes)  Netflix DL (Bytes)  \
0             1.585461e+07        2.501332e+06        8.198936e+06   
1             2.024740e+07        1.911173e+07        1.833841e+07   
2             1.972566e+07        1.469958e+07        1.758779e+07   
3             2.138812e+07        1.514664e+07        1.399465e+07   
4             1.525938e+07        1.896287e+07        1.712458e+07   
...                    ...                 ...                 ...   
149996        1.619167e+07        1.176343e+07        1.788370e+07   
149997        1.387723e+07        8.288284e+06        1.935015e+07   
149998        2.266051e+07        1.855903e+06        9.963942e+06   
149999        8.817106e+06        8.305402e+06        3.322253e+06   
150000        1.163407e+07        1.100941e+07        1.162685e+07   

        Netflix UL (Bytes)  Gaming DL (Bytes)  Gaming UL (Bytes)  \
0             9.656251e+06       2.780823e+08       1.434415e+07   
1             1.722713e+07       6.087501e+08       1.170709e+06   
2             6.163408e+06       2.295846e+08       3.956300e+05   
3             1.097942e+06       7.995382e+08       1.084972e+07   
4             4.152180e+05       5.277072e+08       3.529801e+06   
...                    ...                ...                ...   
149996        1.967816e+07       5.266097e+08       9.197207e+06   
149997        2.129315e+07       6.268931e+08       4.735033e+06   
149998        5.065760e+06       5.535395e+08       1.339432e+07   
149999        1.317259e+07       3.525370e+08       2.529475e+06   
150000        1.100175e+07       4.220447e+08       8.288398e+06   

        Other DL (Bytes)  Othe

#### Task 1.1 - Aggregate the Number of xDR sessions,  Session duration,  the total download (DL) and upload (UL) data , the total data volume (in Bytes) during this session for each application (Social Media, Google, Email, YouTube, Netflix, Gaming).

#### Aggregate the Number of xDR sessions

In [102]:
df.groupby('MSISDN/Number')['Bearer Id'].count()

MSISDN/Number
3.360100e+10    1
3.360100e+10    1
3.360100e+10    1
3.360101e+10    1
3.360101e+10    2
               ..
3.379000e+10    1
3.379000e+10    1
3.197021e+12    1
3.370000e+14    1
8.823971e+14    1
Name: Bearer Id, Length: 106856, dtype: int64

#### Aggregate the Session duration

In [103]:
df.groupby('MSISDN/Number')['Dur. (ms)'].sum()

MSISDN/Number
3.360100e+10    116720.0
3.360100e+10    181230.0
3.360100e+10    134969.0
3.360101e+10     49878.0
3.360101e+10     37104.0
                  ...   
3.379000e+10      8810.0
3.379000e+10    140988.0
3.197021e+12    877385.0
3.370000e+14    253030.0
8.823971e+14    869844.0
Name: Dur. (ms), Length: 106856, dtype: float64

#### Aggregate the total download (DL) and upload (UL) data

In [104]:
df.groupby('MSISDN/Number')['Total DL (Bytes)','Total UL (Bytes)'].sum()

C:\Users\Tijesunimi2-PC\anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


Total DL (Bytes)  Total UL (Bytes)
MSISDN/Number                                    
3.360100e+10       8.426375e+08        36053108.0
3.360100e+10       1.207552e+08        36104459.0
3.360100e+10       5.566597e+08        39306820.0
3.360101e+10       4.019932e+08        20327526.0
3.360101e+10       1.363130e+09        94280527.0
...                         ...               ...
3.379000e+10       6.879252e+08        26716429.0
3.379000e+10       4.445751e+08        35732243.0
3.197021e+12       1.948281e+08        37295915.0
3.370000e+14       5.396350e+08        56652839.0
8.823971e+14       7.869760e+07        60456049.0

[106856 rows x 2 columns]

#### Aggregate the total data volume (in Bytes) during this session for each application (Social Media, Google, Email, YouTube, Netflix, Gaming).

In [105]:
import numpy as np
df.groupby('MSISDN/Number')['Social Media DL (Bytes)','Social Media UL (Bytes)', 'Google DL (Bytes)', 'Google UL (Bytes)',
                            'Email DL (Bytes)', 'Email UL (Bytes)', 'Youtube DL (Bytes)', 'Youtube UL (Bytes)',
                            'Netflix DL (Bytes)','Netflix UL (Bytes)', 'Gaming DL (Bytes)', 'Gaming UL (Bytes)'].sum()


C:\Users\Tijesunimi2-PC\anaconda3\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  


Social Media DL (Bytes)  Social Media UL (Bytes)  \
MSISDN/Number                                                     
3.360100e+10                 2206504.0                  25631.0   
3.360100e+10                 2598548.0                  62017.0   
3.360100e+10                 3148004.0                  47619.0   
3.360101e+10                  251469.0                  28825.0   
3.360101e+10                 2861230.0                  51312.0   
...                                ...                      ...   
3.379000e+10                  234320.0                  65863.0   
3.379000e+10                  442214.0                  56355.0   
3.197021e+12                  668596.0                  46628.0   
3.370000e+14                  496337.0                  25229.0   
8.823971e+14                 1500145.0                  45943.0   

               Google DL (Bytes)  Google UL (Bytes)  Email DL (Bytes)  \
MSISDN/Number                                                           
3.360100e+10           3337123.0          1051882.0          837400.0   
3.360100e+10           4197697.0          1137166.0         2828821.0   
3.360100e+10           3343483.0            99643.0         2436500.0   
3.360101e+10           5937765.0          3740728.0         2178618.0   
3.360101e+10          13728668.0          4770948.0         2247808.0   
...                          ...                ...               ...   
3.379000e+10           6834178.0           697091.0          480946.0   
3.379000e+10           1472406.0          3957299.0         2513433.0   
3.197021e+12           8572779.0          1865881.0          842279.0   
3.370000e+14           8215537.0          1615080.0         2989663.0   
8.823971e+14           5985089.0          3233558.0         2518425.0   

               Email UL (Bytes)  Youtube DL (Bytes)  Youtube UL (Bytes)  \
MSISDN/Number                                                             
3.360100e+10           493962.0          14900201.0           6724347.0   
3.360100e+10           478960.0           5324251.0           7107972.0   
3.360100e+10           768880.0           2137272.0          19196298.0   
3.360101e+10           106052.0           4393123.0           2584198.0   
3.360101e+10          1057661.0          10339971.0          31193031.0   
...                         ...                 ...                 ...   
3.379000e+10           525969.0           8294310.0          18353533.0   
3.379000e+10              664.0           5596862.0          14254710.0   
3.197021e+12           678492.0           9839889.0           2120016.0   
3.370000e+14           328919.0          16690728.0          20044212.0   
8.823971e+14           812549.0          18980320.0          21960390.0   

               Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
MSISDN/Number                                                              
3.360100e+10           10265105.0          16915876.0       8.110911e+08   
3.360100e+10             770569.0          10451194.0       1.050353e+08   
3.360100e+10           16525919.0           2827981.0       5.290685e+08   
3.360101e+10            1157362.0            784730.0       3.880748e+08   
3.360101e+10           24971647.0          24230077.0       1.308981e+09   
...                           ...                 ...                ...   
3.379000e+10           14754741.0            147797.0       6.573267e+08   
3.379000e+10            6929961.0           1601099.0       4.276202e+08   
3.197021e+12           10340908.0          16251392.0       1.645636e+08   
3.370000e+14            9983678.0          20921364.0       5.012590e+08   
8.823971e+14           21621336.0           7224894.0       2.809228e+07   

               Gaming UL (Bytes)  
MSISDN/Number                     
3.360100e+10           1367528.0  
3.360100e+10          14714780.0  
3.360100e+10           9759228.0  
3.360101e+10           3051292.0  
3.360101e+10      

#### Start by identifying the top 10 handsets used by the customers.

In [97]:
df.groupby('Handset Type')['MSISDN/Number'].count().sort_values(ascending=False).head(10)

Handset Type
Huawei B528S-23A                19727
Apple iPhone 6S (A1688)          9413
Apple iPhone 6 (A1586)           9012
undefined                        8931
Apple iPhone 7 (A1778)           6304
Apple iPhone Se (A1723)          5176
Apple iPhone 8 (A1905)           4985
Apple iPhone Xr (A2105)          4562
Samsung Galaxy S8 (Sm-G950F)     4480
Apple iPhone X (A1901)           3810
Name: MSISDN/Number, dtype: int64

#### Then, identify the top 3 handset manufacturers

In [98]:
df.groupby(['Handset Manufacturer'])['MSISDN/Number'].count().sort_values(ascending=False).head(3)

Handset Manufacturer
Apple      59464
Samsung    40579
Huawei     34366
Name: MSISDN/Number, dtype: int64

#### Next, identify the top 5 handsets per handset manufacturer

In [99]:
df.groupby(['Handset Manufacturer','Handset Type'])['MSISDN/Number'].count().sort_values(ascending=False).head(9)

Handset Manufacturer  Handset Type                
Huawei                Huawei B528S-23A                19727
Apple                 Apple iPhone 6S (A1688)          9413
                      Apple iPhone 6 (A1586)           9012
undefined             undefined                        8931
Apple                 Apple iPhone 7 (A1778)           6304
                      Apple iPhone Se (A1723)          5176
                      Apple iPhone 8 (A1905)           4985
                      Apple iPhone Xr (A2105)          4562
Samsung               Samsung Galaxy S8 (Sm-G950F)     4480
Name: MSISDN/Number, dtype: int64

#### Make a short interpretation and recommendation to marketing teams

#### Task 1.2 - Conduct an exploratory data analysis on those data & communicate useful insights. Ensure that you identify and treat all missing values and outliers in the dataset by replacing by the mean of the corresponding column.